### DISCUSSION TITLE
Using drift without flutter

### DISCUSSION THREAD

**Discussion by bmerchant22** [simolus3/drift#3028] [Q&A]
Hello everyone. I am using drift in my backend project in dart without flutter. But, it's someway using flutter libraries, and when I run my dart code, it gives errors. Can you please tell me how to use drift in dart without flutter.

**Comment by simolus3** [comment]
That's awesome, I'm always excited to see Dart being used outside of Flutter! Drift was designed to support a pure Dart environment, but admittedly the "getting started" docs are kinds of Flutter-centric since that's realistically what almost all users need.
The docs suggest two Flutter-only packages which you need to avoid in backend projects, `sqlite3_flutter_libs` and `path_provider`. `path_provider` is not hard to avoid, you just need to find your own logic for picking the location of the database. `sqlite3_flutter_libs` is responsible for linking a recent sqlite3 build with Flutter applications. Drift will fall back to the system's sqlite installation though. You can also see [this](https://pub.dev/packages/sqlite3#manually-providing-sqlite3-libraries) for more details and ideas on how to include your own sqlite3 copy. Drift also [supports postgres](https://drift.simonbinder.eu/docs/platforms/postgres/) for which you don't need either package so that could be interesting as well for backends.

**Reply by bmerchant22** [reply]
Actually, I am developing a mail client, and I want to store mails in local storage of the users. And I want it to be used on all platforms i.e. android, IOS, windows and linux. So, if I use postgreSQL, will it work on all platforms?

**Comment by bmerchant22** [comment]
This is my `bin/main.dart`

```dart
import 'dart:ffi';
import 'dart:io';
import 'package:path/path.dart';
import 'package:sqlite3/open.dart';
import 'package:sqlite3/sqlite3.dart';

void main() async {
  open.overrideFor(OperatingSystem.windows, _openOnWindows);

  final db = sqlite3.openInMemory();

  db.dispose();
}

DynamicLibrary _openOnWindows() {
  final scriptDir = File(Platform.script.toFilePath()).parent;
  final libraryNextToScript = File(join(scriptDir.path, 'sqlite3.dll'));
  return DynamicLibrary.open(libraryNextToScript.path);
}
```

And this is my `lib/src/db.dart`

```dart
import 'dart:ffi';
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:sqlite3/open.dart';
import 'package:sqlite3/sqlite3.dart';

part 'db.g.dart';

class MailOverview extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get subject => text().withLength(min: 1, max: 255)();
  TextColumn get sender => text().withLength(min: 1, max: 255)();
  DateTimeColumn get date => dateTime()();
}

@DriftDatabase(tables: [MailOverview])
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;
}

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    final dbFolder = new Directory('mails');
    final file = File('${dbFolder.path}/db.sqlite');

    return NativeDatabase.createInBackground(file);
  });
}
```

I am getting error with NativeDatabase class, its recognizing this class, what should I do??

**Reply by simolus3** [reply]
> I am getting error with NativeDatabase class, its recognizing this class, what should I do??

You need to import `package:drift/native.dart`.

**Comment by bmerchant22** [comment]
I am somewhat unclear about the sqlite3.so or .dll file. I don't have that much experience with sqlite. Can you tell me where to find this file and how to configure it for all the devices i.e. android, linux, windows etc. Also, why is it necessary to override and use a custom sqlite3.so file.

**Reply by simolus3** [reply]
sqlite is a C library. To access it from Dart, it needs to be available in the form of a "dynamic library" which contains the compiled C code for sqlite3 and addresses for the exposed functions which we're calling from Dart.

On Android and iOS you're not doing anything without Flutter either way, so just depend on `sqlite3_flutter_libs` which includes Gradle/CocoaPods buildscripts to include the library you need with your app. The [sqlite3 docs](https://pub.dev/packages/sqlite3#supported-platforms) contain some pointers on how to get  sqlite3 on other platforms.
The easiest approach is to use the sqlite3 from a package manager or the operating system. On macOS, sqlite3 comes with the OS for instance and you don't really need to do anything. On Debian you can install `libsqlite3-dev` and virtually every other Linux distro will have a similar package containing a `libsqlite3.so`. Windows is tricky since there's no universal package manager, but you can download a DLL [here](https://sqlite.org/download.html) ("Precompiled Binaries for Windows"). You also have the option of shipping sqlite3 with your app for Linux and macOS by either downloading a prebuilt binary or compiling it yourself (sqlite3 provides plenty of docs on how to do that).
